luda-adventure/src/app/game/types/game_object/types/positioner/positioner.rs

```diff
@@ -50,107 +51,116 @@ impl Positioner {
+                    &character_collision_box,
+                    &target_collision_box_list,
+                    prediction_start_time,
+                    directed_movement.end_time,
+                    prediction_start_position,
+                ) {
+                    let movement_until_collision = Movement {
+                        start_time: prediction_start_time,
+                        end_time: collision.start_time,
+                        start_position: prediction_start_position,
+                        end_position: collision.start_position,
+                        velocity: velocity_after_last_collision,
+                        movement_state: MovementState::MoveAlongAxisToCollide,
+                    };
+                    self.movement_plan
+                        .predicted_movement_list
+                        .push(movement_until_collision);
+                } else {
+                    let delta_time = last_collision.end_time - last_collision.start_time;
+                    let end_position =
+                        velocity_after_last_collision * delta_time + prediction_start_position;
+                    let movement_until_last_collision_end = Movement {
+                        start_time: prediction_start_time,
+                        end_time: last_collision.end_time,
+                        start_position: prediction_start_position,
+                        velocity: velocity_after_last_collision,
+                        end_position,
+                        movement_state: MovementState::MoveAlongAxis,
+                    };
+                    self.movement_plan
+                        .predicted_movement_list
+                        .push(movement_until_last_collision_end);
                 }
             }
-            None => {
-                let movement_until_end = Movement {
-                    start_time: prediction_start_time,
-                    end_time: self.movement_plan.directed_movement.end_time,
-                    start_position: prediction_start_position,
-                    velocity: self.movement_plan.directed_movement.velocity,
-                    end_position: self.movement_plan.directed_movement.end_position,
-                };
+            super::MovementState::MoveAlongAxisToCollide | super::MovementState::Stuck => {
+                let stay_forever =
+                    Movement::stay_forever(prediction_start_position, prediction_start_time);
                 self.movement_plan
                     .predicted_movement_list
-                    .push(movement_until_end);
+                    .push(stay_forever);
             }
-        };
+        }
     }
 
     pub fn get_position(&self, current_time: Time) -> Xy<crate::app::game::Tile> {
         self.movement_plan
             .get_position(current_time)
             .expect("get_position() of PlayerCharacter called before prediction")
     }
+
+    fn get_last_prediction(&self) -> Movement {
```

Comment:
FYI: You can return reference or impl Copy for Movement

Reply:
[6982da7](https://github.com/NamseEnt/namseent/pull/359/commits/6982da7181a4446bd46f8e1e029a8020f267c7cb)